core/src/main/java/be/orbinson/aem/dictionarytranslator/servlets/action/CreateMessageEntryServlet.java

```diff
@@ -32,55 +29,48 @@
 )
 public class CreateMessageEntryServlet extends AbstractDictionaryServlet {
 
+    public CreateMessageEntryServlet() {
+        super("Unable to create message entry");
+    }
+
     private static final Logger LOG = LoggerFactory.getLogger(CreateMessageEntryServlet.class);
 
     @Reference
     private transient DictionaryService dictionaryService;
 
     @Override
-    protected void doPost(SlingHttpServletRequest request, @NotNull SlingHttpServletResponse response) throws IOException {
+    protected void internalDoPost(SlingHttpServletRequest request, HtmlResponse htmlResponse) throws Throwable {
         String key = getMandatoryParameter(request, "key", false);
         String dictionary = getMandatoryParameter(request, "dictionary", false);
-
-        createMessageEntry(request, response, dictionary, key);
-        
+        htmlResponse.setCreateRequest(true);
+        createMessageEntry(request, htmlResponse, dictionary, key);
     }
 
-    private void createMessageEntry(SlingHttpServletRequest request, @NotNull SlingHttpServletResponse response, String path, String key) throws IOException {
+    private void createMessageEntry(SlingHttpServletRequest request, HtmlResponse htmlResponse, String path, String key) throws IOException, DictionaryException {
         ResourceResolver resourceResolver = request.getResourceResolver();
-        try {
-            Collection<Dictionary> dictionaries = dictionaryService.getDictionaries(resourceResolver, path);
-            if (dictionaries.isEmpty()) {
-                HtmlResponse htmlResponse = new HtmlResponse();
-                htmlResponse.setStatus(HttpServletResponse.SC_BAD_REQUEST, String.format("No dictionaries found at path '%s'", path));
-                htmlResponse.send(response, true);
-                return;
-            }
-            for (Dictionary languageDictionary : dictionaries) {
-                if (!languageDictionary.getEntries().containsKey(key)) {
-                    String name = languageDictionary.getLanguage().toLanguageTag();
-                    String message = getOptionalParameter(request, name, true).orElse("");
-                    boolean useEmpty = getOptionalParameter(request, name + CombiningMessageEntryDatasourceForDialog.SUFFIX_USE_EMPTY, false, Boolean::parseBoolean).orElse(false);
-                    Optional<String> messageOptional;
-                    if (useEmpty || !message.isEmpty()) {
-                        messageOptional = Optional.of(message);
-                    } else {
-                        messageOptional = Optional.empty();
-                    }
-                    languageDictionary.createEntry(resourceResolver, key, messageOptional);
+        Collection<Dictionary> dictionaries = dictionaryService.getDictionaries(resourceResolver, path);
+        if (dictionaries.isEmpty()) {
+            htmlResponse.setStatus(HttpServletResponse.SC_BAD_REQUEST, String.format("No dictionaries found at path '%s'", path));
```

Comment:
404?

Reply:
If we want to expose the API's as real REST API's => 404. 
If they are tightly bound with our granite ui, then it means someone is doing something fishy and shouldn't be in this situation => 400

So for me both are fine :) I don't think anyone would use these API's outside the framework but you never know

Reply:
I agree that it shouldn't matter for this case, because the UI exposed error looks the same for both 404 and 400 (and every other non-success status code). However this situation does not necessarily mean someone is doing something fishy.
We are in a heavy concurrent system where arbitrarily many users could potentially modify dictionary resources via multiple means:
- installing content packages covering dictionaries
- working with AEM Dictionary Translator
- working with Adobe's Translator
- working with the Translation Project

Those concurrent situations may easily lead to situations where every client behaves correctly, but still one gets a 404/400 because the underlying repository was meanhwhile removed by someone else.

Reply:
True 👍